service/history/api/multioperation/api.go

```diff
@@ -0,0 +1,167 @@
+	visibilityManager manager.VisibilityManager,
+	matchingClient matchingservice.MatchingServiceClient,
+) (_ *historyservice.ExecuteMultiOperationResponse, retError error) {
+	namespaceEntry, err := api.GetActiveNamespace(shardContext, namespace.ID(req.GetNamespaceId()))
+	if err != nil {
+		return nil, err
+	}
+	namespaceID := namespaceEntry.ID()
+
+	startReq := req.Operations[0].GetStartWorkflow()
+	updateReq := req.Operations[1].GetUpdateWorkflow()
+
+	// grab current Workflow lease, if there is one
+	currentWorkflowLease, err := workflowConsistencyChecker.GetWorkflowLease(
+		ctx,
+		nil,
+		api.BypassMutableStateConsistencyPredicate,
+		definition.NewWorkflowKey(
+			string(namespaceID),
+			startReq.StartRequest.WorkflowId,
+			"",
+		),
+		workflow.LockPriorityHigh,
+	)
+	switch err.(type) {
+	case nil:
+		defer func() { currentWorkflowLease.GetReleaseFn()(retError) }()
+	case *serviceerror.NotFound:
+		currentWorkflowLease = nil
+	default:
+		return nil, err
+	}
+
+	if currentWorkflowLease != nil {
+		return nil, serviceerror.NewUnimplemented("not implemented yet")
+	}
+
+	starter, err := startworkflow.NewStarter(
+		shardContext,
+		workflowConsistencyChecker,
+		tokenSerializer,
+		visibilityManager,
+		startReq,
+	)
+	if err != nil {
+		// TODO: send per-operation error details
+		return nil, err
+	}
+
+	var workflowCtx workflow.Context
+	var updateErr error
+	updater := updateworkflow.NewUpdater(
+		shardContext,
+		workflowConsistencyChecker,
+		matchingClient,
+		updateReq,
+	)
+	startResp, err := starter.Invoke(
+		ctx,
+		func(lease api.WorkflowLease) error {
+			workflowCtx = lease.GetContext()
+			_, updateErr = updater.ApplyRequest(ctx, lease)
+			return updateErr
+		},
+	)
+	if err != nil {
+		// TODO: send per-operation error details
+		return nil, err
+	}
+	startOpResp := &historyservice.ExecuteMultiOperationResponse_Response{
+		Response: &historyservice.ExecuteMultiOperationResponse_Response_StartWorkflow{
+			StartWorkflow: startResp,
+		},
+	}
+
+	if updateErr != nil {
+		updateResp := updater.OnError(updateErr)
+		return &historyservice.ExecuteMultiOperationResponse{
+			Responses: []*historyservice.ExecuteMultiOperationResponse_Response{
+				startOpResp,
+				{
+					Response: &historyservice.ExecuteMultiOperationResponse_Response_UpdateWorkflow{
+						UpdateWorkflow: updateResp,
+					},
+				},
+			},
+		}, nil
+	}
+
+	// Without this, there's no Update registry on the call from Matching back to History.
+	// TODO: eventually, we'll want to put the MS into the cache as well
+	if _, err = workflowConsistencyChecker.GetWorkflowCache().Put(
+		shardContext,
+		namespaceID,
+		&commonpb.WorkflowExecution{WorkflowId: req.WorkflowId, RunId: startResp.RunId},
+		workflowCtx,
+		shardContext.GetMetricsHandler(),
+	); err != nil {
+		return nil, err
+	}
+
+	updateResp, err := updater.OnSuccess(ctx)
+	if err != nil {
+		// TODO: send Update outcome failure instead
+		return nil, err
```

Comment:
Please add some context to the failure here; there are a number of stages here and it would be good to know at which one we failed